proposals/NNNN-task-executor-preference.md

```diff
@@ -0,0 +1,664 @@
+} 
+```
+
+or, for a specific scope using the `withTaskExecutor` method:
+
+```swift
+await withTaskExecutor(executor) { 
+  // if not already running on specified 'executor'
+  // the withTaskExecutor would hop to it, and run this closure on it.
+
+  // task groups
+  await withDiscardingTaskGroup { group in 
+    group.addTask {
+      // starts and runs on the 'executor'
+      await nonisolatedAsyncFunc()
+    }
+  }
+  
+  // async let 
+  async let number = nonisolatedAsyncFunc() // starts and runs on 'executor'
+  await number
+}
+```
+
+Notably, the task executor preference is in effect for the entire structured task hierarchy while running in a task or scope where a task executor preference is set. For example, the following snippet illustrates child tasks created inside of a `withTaskExecutor`.
+
+If a task with such executor preference encounters code which is `isolated` to some specific actor, it would adhere to that requirement and hop to it as expected:
+
+```swift
+let capy: Capybara = Capybara()
+actor Capybara { func eat() {} } 
+
+Task(on: executor) {
+  // starts on 'executor', however...
+  try await capy.eat() // still executes actor isolated code on the actor's executor, as expected
+}
+```
+
+## Detailed design
+
+### Setting task executor preference
+
+A new concept of task executor preference is added to Swift Concurrency tasks. This preference is stored in a task and propagated throughout child tasks (such as ones created by TaskGroups and async let).
+
+The preference can be set using various APIs that will be discussed in detail in their respective sections. The first of those APIs is `withTaskExecutor` which is can be called inside an asynchronous context to both ensure we’re executing on the expected executor, as well as set the task executor preference for the duration of the operation closure:
+
+```swift
+await withTaskExecutor(someExecutor) { 
+ // guaranteed to be executing on someExecutor
+}
+```
+
+Once set, the effect of an executor preference is such that a nonisolated func instead of immediately hopping to the global pool, it may hop to the preferred executor, e.g.:
+
+```swift
+nonisolated func tryMe() async { 
+ // ...
+}
+
+let preferredExecutor: SomeConcreteExecutor = ...
+Task(on: preferredExecutor) { 
+  preferredExecutor.assertIsolated()
+  await tryMe() // tryMe body would execute on 'preferredExecutor'
+}
+
+ await tryMe() // tryMe body would execute on 'default global concurrent executor'
+```
+
+### Task executor preference inheritance in Structured Concurrency
+
+Task executor preference is inherited by child tasks and actors which do not declare an explicit executor (so-called "default actors"), and is *not* inherited by un-structured tasks. 
+
+Specifically:
+
+* **Do** inherit task executor preference
+    * TaskGroup’s `addTask()`, unless overridden with explicit parameter
+    * `async let`
+    * methods on actors which which do not declare an explicit `unownedExecutor` requirement
```

Comment:
This reads as if methods can declare an `unownedExecutor` requirement. Can we change the wording a bit.